File: crates/trendlab-core/src/metrics.rs
//! Performance metrics calculations.

use crate::backtest::{BacktestResult, Trade};
use serde::{Deserialize, Deserializer, Serialize};

/// Deserialize a field that may be null as the default value.
fn deserialize_null_as_default<'de, D, T>(deserializer: D) -> Result<T, D::Error>
where
    D: Deserializer<'de>,
    T: Default + Deserialize<'de>,
{
    let opt = Option::deserialize(deserializer)?;
    Ok(opt.unwrap_or_default())
}

/// Performance metrics for a backtest run.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct Metrics {
    /// Total return (as decimal, e.g., 0.25 = 25%)
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub total_return: f64,

    /// Compound annual growth rate
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub cagr: f64,

    /// Annualized Sharpe ratio (assuming 252 trading days)
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub sharpe: f64,

    /// Annualized Sortino ratio
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub sortino: f64,

    /// Maximum drawdown (as a positive percentage, e.g., 0.20 = 20%)
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub max_drawdown: f64,

    /// Calmar ratio (CAGR / Max Drawdown)
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub calmar: f64,

    /// Win rate (winning trades / total trades)
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub win_rate: f64,

    /// Profit factor (gross profit / gross loss)
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub profit_factor: f64,

    /// Total number of trades
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub num_trades: u32,

    /// Annual turnover (as multiple of capital)
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub turnover: f64,

    /// Maximum consecutive losing trades
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub max_consecutive_losses: u32,

    /// Maximum consecutive winning trades
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub max_consecutive_wins: u32,

    /// Average length of losing streaks
    #[serde(default, deserialize_with = "deserialize_null_as_default")]
    pub avg_losing_streak: f64,
}


/// Compute all metrics from a BacktestResult.
pub fn compute_metrics(result: &BacktestResult, initial_cash: f64) -> Metrics {
    if result.equity.is_empty() {
        return Metrics::default();
    }

    let equity_curve: Vec<f64> = result.equity.iter().map(|e| e.equity).collect();
    let last_equity = equity_curve.last().copied().unwrap_or(initial_cash);

    // Total return
    let total_return = if initial_cash > 0.0 {
        (last_equity - initial_cash) / initial_cash
    } else {
        0.0
    };

    // Calculate years from first to last bar
    let years = if result.equity.len() >= 2 {
        let first_ts = result.equity.first().unwrap().ts;
        let last_ts = result.equity.last().unwrap().ts;
        let duration = last_ts.signed_duration_since(first_ts);
        duration.num_days() as f64 / 365.25
    } else {
        0.0
    };

    // CAGR
    let cagr = calculate_cagr(initial_cash, last_equity, years);

    // Max drawdown
    let max_drawdown = calculate_max_drawdown(&equity_curve);

    // Calmar ratio (CAGR / Max Drawdown)
    let calmar = if max_drawdown > 0.0 {
        cagr / max_drawdown
    } else {
        0.0
    };

    // Daily returns for Sharpe calculation
    // Guard against division by zero if equity ever hits zero
    let daily_returns: Vec<f64> = equity_curve
        .windows(2)
        .map(|w| {
            if w[0].abs() > 1e-10 {
                (w[1] - w[0]) / w[0]
            } else {
                0.0
            }
        })
        .collect();

    let sharpe = calculate_sharpe(&daily_returns);
    let sortino = calculate_sortino(&daily_returns);

    // Trade-based metrics
    let num_trades = result.trades.len() as u32;
    let winning_trades = result.trades.iter().filter(|t| t.net_pnl > 0.0).count();
    let win_rate = if num_trades > 0 {
        winning_trades as f64 / num_trades as f64
    } else {
        0.0
    };

    // Profit factor
    let gross_profit: f64 = result
        .trades
        .iter()
        .filter(|t| t.net_pnl > 0.0)
        .map(|t| t.net_pnl)
        .sum();
    let gross_loss: f64 = result
        .trades
        .iter()
        .filter(|t| t.net_pnl < 0.0)
        .map(|t| t.net_pnl.abs())
        .sum();
    // Profit factor: cap at 999.99 when gross_loss is 0 to avoid JSON null (INFINITY → null)
    let profit_factor = if gross_loss > 0.0 {
        (gross_profit / gross_loss).min(999.99)
    } else if gross_profit > 0.0 {
        999.99 // Capped: no losses means "infinite" profit factor
    } else {
        0.0
    };

    // Turnover (total traded notional / average capital)
    let total_traded: f64 = result.fills.iter().map(|f| (f.qty * f.price).abs()).sum();
    let avg_capital = (initial_cash + last_equity) / 2.0;
    let turnover = if years > 0.0 && avg_capital > 0.0 {
        (total_traded / avg_capital) / years
    } else {
        0.0
    };

    // Consecutive streak metrics
    let (max_consecutive_losses, max_consecutive_wins, avg_losing_streak) =
        calculate_streaks(&result.trades);

    Metrics {
        total_return,
        cagr,
        sharpe,
        sortino,
        max_drawdown,
        calmar,
        win_rate,
        profit_factor,
        num_trades,
        turnover,
        max_consecutive_losses,
        max_consecutive_wins,
        avg_losing_streak,
    }
}

/// Calculate CAGR from initial and final values over a number of years.
pub fn calculate_cagr(initial: f64, final_value: f64, years: f64) -> f64 {
    if initial <= 0.0 || years <= 0.0 {
        return 0.0;
    }
    (final_value / initial).powf(1.0 / years) - 1.0
}

/// Calculate annualized Sharpe ratio from daily returns.
///
/// Assumes 252 trading days per year and risk-free rate of 0.
pub fn calculate_sharpe(daily_returns: &[f64]) -> f64 {
    if daily_returns.is_empty() {
        return 0.0;
    }

    let n = daily_returns.len() as f64;
    let mean = daily_returns.iter().sum::<f64>() / n;
    let variance = daily_returns
        .iter()
        .map(|r| (r - mean).powi(2))
        .sum::<f64>()
        / n;
    let std_dev = variance.sqrt();

    if std_dev == 0.0 {
        return 0.0;
    }

    // Annualize: multiply mean by 252, std by sqrt(252)
    (mean * 252.0) / (std_dev * 252.0_f64.sqrt())
}

/// Calculate annualized Sortino ratio from daily returns.
///
/// Like Sharpe but only penalizes downside volatility.
pub fn calculate_sortino(daily_returns: &[f64]) -> f64 {
    if daily_returns.is_empty() {
        return 0.0;
    }

    let n = daily_returns.len() as f64;
    let mean = daily_returns.iter().sum::<f64>() / n;

    // Downside deviation: only consider returns below zero
    let downside_variance = daily_returns
        .iter()
        .map(|r| if *r < 0.0 { r.powi(2) } else { 0.0 })
        .sum::<f64>()
        / n;
    let downside_dev = downside_variance.sqrt();

    if downside_dev == 0.0 {
        return 0.0;
    }

    // Annualize
    (mean * 252.0) / (downside_dev * 252.0_f64.sqrt())
}

/// Calculate maximum drawdown from an equity curve.
pub fn calculate_max_drawdown(equity_curve: &[f64]) -> f64 {
    if equity_curve.is_empty() {
        return 0.0;
    }

    let mut peak = equity_curve[0];
    let mut max_dd = 0.0;

    for &equity in equity_curve {
        if equity > peak {
            peak = equity;
        }
        // Guard against division by zero if peak is somehow zero
        let dd = if peak.abs() > 1e-10 {
            (peak - equity) / peak
        } else {
            0.0
        };
        if dd > max_dd {
            max_dd = dd;
        }
    }

    max_dd
}

/// Calculate consecutive win/loss streaks from trades.
///
/// Returns: (max_consecutive_losses, max_consecutive_wins, avg_losing_streak)
pub fn calculate_streaks(trades: &[Trade]) -> (u32, u32, f64) {
    if trades.is_empty() {
        return (0, 0, 0.0);
    }

    let mut max_losses = 0u32;
    let mut max_wins = 0u32;
    let mut current_losses = 0u32;
    let mut current_wins = 0u32;

    // Track all losing streaks for average calculation
    let mut losing_streaks: Vec<u32> = Vec::new();

    for trade in trades {
        if trade.net_pnl > 0.0 {
            // Winner
            current_wins += 1;
            if current_losses > 0 {
                losing_streaks.push(current_losses);
                max_losses = max_losses.max(current_losses);
                current_losses = 0;
            }
        } else {
            // Loser (including breakeven as loss for conservative counting)
            current_losses += 1;
            if current_wins > 0 {
                max_wins = max_wins.max(current_wins);
                current_wins = 0;
            }
        }
    }

    // Handle final streak
    if current_losses > 0 {
        losing_streaks.push(current_losses);
        max_losses = max_losses.max(current_losses);
    }
    if current_wins > 0 {
        max_wins = max_wins.max(current_wins);
    }

    // Calculate average losing streak
    let avg_losing_streak = if losing_streaks.is_empty() {
        0.0
    } else {
        losing_streaks.iter().sum::<u32>() as f64 / losing_streaks.len() as f64
    };

    (max_losses, max_wins, avg_losing_streak)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::backtest::{Fill, Side, TradeDirection};
    use chrono::Utc;

    /// Helper to create a trade with just net_pnl (for streak testing)
    fn make_trade(net_pnl: f64) -> Trade {
        let fill = Fill {
            ts: Utc::now(),
            side: Side::Buy,
            qty: 1.0,
            price: 100.0,
            fees: 0.0,
            raw_price: 100.0,
            atr_at_fill: None,
        };
        Trade {
            entry: fill.clone(),
            exit: fill,
            gross_pnl: net_pnl,
            net_pnl,
            direction: TradeDirection::Long,
        }
    }

    #[test]
    fn test_cagr() {
        // $100 -> $200 in 5 years = ~14.87% CAGR
        let cagr = calculate_cagr(100.0, 200.0, 5.0);
        assert!((cagr - 0.1487).abs() < 0.001);
    }

    #[test]
    fn test_max_drawdown() {
        let equity = vec![100.0, 110.0, 105.0, 120.0, 90.0, 100.0];
        let dd = calculate_max_drawdown(&equity);
        // Peak was 120, trough was 90 -> 25% drawdown
        assert!((dd - 0.25).abs() < 0.001);
    }

    #[test]
    fn test_streaks_empty() {
        let trades: Vec<Trade> = vec![];
        let (max_losses, max_wins, avg_losing) = calculate_streaks(&trades);
        assert_eq!(max_losses, 0);
        assert_eq!(max_wins, 0);
        assert!((avg_losing - 0.0).abs() < 0.001);
    }

    #[test]
    fn test_streaks_all_winners() {
        let trades = vec![make_trade(100.0), make_trade(50.0), make_trade(200.0)];
        let (max_losses, max_wins, avg_losing) = calculate_streaks(&trades);
        assert_eq!(max_losses, 0);
        assert_eq!(max_wins, 3);
        assert!((avg_losing - 0.0).abs() < 0.001);
    }

    #[test]
    fn test_streaks_all_losers() {
        let trades = vec![make_trade(-100.0), make_trade(-50.0), make_trade(-200.0)];
        let (max_losses, max_wins, avg_losing) = calculate_streaks(&trades);
        assert_eq!(max_losses, 3);
        assert_eq!(max_wins, 0);
        assert!((avg_losing - 3.0).abs() < 0.001); // One streak of 3
    }

    #[test]
    fn test_streaks_alternating() {
        // W, L, W, L, W, L
        let trades = vec![
            make_trade(100.0),
            make_trade(-50.0),
            make_trade(100.0),
            make_trade(-50.0),
            make_trade(100.0),
            make_trade(-50.0),
        ];
        let (max_losses, max_wins, avg_losing) = calculate_streaks(&trades);
        assert_eq!(max_losses, 1);
        assert_eq!(max_wins, 1);
        assert!((avg_losing - 1.0).abs() < 0.001); // Three streaks of 1
    }

    #[test]
    fn test_streaks_mixed() {
        // W, W, L, L, L, W, L, L
        let trades = vec![
            make_trade(100.0),  // W
            make_trade(50.0),   // W (2 win streak)
            make_trade(-100.0), // L
            make_trade(-50.0),  // L
            make_trade(-200.0), // L (3 loss streak)
            make_trade(150.0),  // W
            make_trade(-75.0),  // L
            make_trade(-25.0),  // L (2 loss streak)
        ];
        let (max_losses, max_wins, avg_losing) = calculate_streaks(&trades);
        assert_eq!(max_losses, 3);
        assert_eq!(max_wins, 2);
        // Two losing streaks: 3 and 2, average = 2.5
        assert!((avg_losing - 2.5).abs() < 0.001);
    }

    #[test]
    fn test_streaks_breakeven_as_loss() {
        // Breakeven (0.0) counts as loss for conservative counting
        let trades = vec![
            make_trade(100.0), // W
            make_trade(0.0),   // L (breakeven)
            make_trade(-50.0), // L
        ];
        let (max_losses, max_wins, _avg_losing) = calculate_streaks(&trades);
        assert_eq!(max_losses, 2); // Breakeven + loss = 2 streak
        assert_eq!(max_wins, 1);
    }
}

File: scripts/gpu_sweep/metrics.py
"""
GPU-accelerated performance metrics with exact Rust parity.

All metrics match the implementations in crates/trendlab-core/src/metrics.rs.
"""

import cupy as cp
import numpy as np


def compute_returns_gpu(equity: cp.ndarray) -> cp.ndarray:
    """
    Compute daily returns from equity curve.

    Args:
        equity: Equity curve, shape [num_configs, num_bars]

    Returns:
        Daily returns, shape [num_configs, num_bars - 1]
        r[t] = (equity[t] - equity[t-1]) / equity[t-1]
    """
    # Avoid division by zero
    prev_equity = equity[:, :-1]
    prev_equity = cp.where(cp.abs(prev_equity) < 1e-10, 1e-10, prev_equity)

    returns = (equity[:, 1:] - equity[:, :-1]) / prev_equity
    return cp.nan_to_num(returns, nan=0.0, posinf=0.0, neginf=0.0)


def sharpe_gpu(equity: cp.ndarray) -> cp.ndarray:
    """
    Compute annualized Sharpe ratio for multiple equity curves.

    Matches Rust calculate_sharpe() exactly:
    - Assumes 252 trading days per year
    - Risk-free rate of 0
    - Uses population standard deviation (n, not n-1)

    Args:
        equity: Equity curve, shape [num_configs, num_bars]

    Returns:
        Sharpe ratios, shape [num_configs]
    """
    if equity.shape[1] < 2:
        return cp.zeros(equity.shape[0], dtype=cp.float32)

    # Daily returns
    returns = compute_returns_gpu(equity)

    # Mean and standard deviation
    n = returns.shape[1]
    mean_ret = cp.sum(returns, axis=1) / n

    # Population variance (matching Rust)
    variance = cp.sum((returns - mean_ret[:, None]) ** 2, axis=1) / n
    std_ret = cp.sqrt(variance)

    # Annualize: (mean * 252) / (std * sqrt(252))
    # This simplifies to: mean * sqrt(252) / std
    sharpe = (mean_ret * 252.0) / (std_ret * cp.sqrt(252.0) + 1e-10)

    return cp.where(std_ret == 0, 0.0, sharpe).astype(cp.float32)


def sortino_gpu(equity: cp.ndarray) -> cp.ndarray:
    """
    Compute annualized Sortino ratio for multiple equity curves.

    Like Sharpe but only penalizes downside volatility.
    Matches Rust calculate_sortino() exactly.

    Args:
        equity: Equity curve, shape [num_configs, num_bars]

    Returns:
        Sortino ratios, shape [num_configs]
    """
    if equity.shape[1] < 2:
        return cp.zeros(equity.shape[0], dtype=cp.float32)

    # Daily returns
    returns = compute_returns_gpu(equity)

    # Mean return
    n = returns.shape[1]
    mean_ret = cp.sum(returns, axis=1) / n

    # Downside deviation: only consider returns below zero
    negative_returns = cp.where(returns < 0, returns, 0.0)
    downside_variance = cp.sum(negative_returns**2, axis=1) / n
    downside_dev = cp.sqrt(downside_variance)

    # Annualize
    sortino = (mean_ret * 252.0) / (downside_dev * cp.sqrt(252.0) + 1e-10)

    return cp.where(downside_dev == 0, 0.0, sortino).astype(cp.float32)


def cagr_gpu(
    equity: cp.ndarray,
    initial_cash: float,
    years: float,
) -> cp.ndarray:
    """
    Compute CAGR for multiple equity curves.

    Matches Rust calculate_cagr() exactly.

    Args:
        equity: Equity curve, shape [num_configs, num_bars]
        initial_cash: Starting capital
        years: Number of years in backtest period

    Returns:
        CAGR values, shape [num_configs]
    """
    if initial_cash <= 0.0 or years <= 0.0:
        return cp.zeros(equity.shape[0], dtype=cp.float32)

    final_equity = equity[:, -1]
    cagr = (final_equity / initial_cash) ** (1.0 / years) - 1.0

    return cagr.astype(cp.float32)


def total_return_gpu(equity: cp.ndarray, initial_cash: float) -> cp.ndarray:
    """
    Compute total return for multiple equity curves.

    Args:
        equity: Equity curve, shape [num_configs, num_bars]
        initial_cash: Starting capital

    Returns:
        Total returns (as decimals), shape [num_configs]
    """
    if initial_cash <= 0.0:
        return cp.zeros(equity.shape[0], dtype=cp.float32)

    final_equity = equity[:, -1]
    return ((final_equity - initial_cash) / initial_cash).astype(cp.float32)


def _running_max_gpu(arr: cp.ndarray) -> cp.ndarray:
    """
    Compute running maximum along axis=1 without cp.maximum.accumulate.

    Uses a parallel prefix-sum style algorithm with log(n) iterations.
    This works by repeatedly computing max with shifted versions.

    Args:
        arr: Input array, shape [num_configs, num_bars]

    Returns:
        Running maximum, shape [num_configs, num_bars]
    """
    n_bars = arr.shape[1]
    if n_bars == 0:
        return arr.copy()

    result = arr.copy()

    # Parallel prefix max using doubling technique
    # Each iteration i computes max with offset 2^i
    offset = 1
    while offset < n_bars:
        # Create shifted version (shift right by offset, pad with -inf)
        shifted = cp.full_like(result, -cp.inf)
        shifted[:, offset:] = result[:, :-offset]
        result = cp.maximum(result, shifted)
        offset *= 2

    return result


def max_drawdown_gpu(equity: cp.ndarray) -> cp.ndarray:
    """
    Compute max drawdown for multiple equity curves.

    Matches Rust calculate_max_drawdown() exactly.
    Uses running peak tracking.

    Args:
        equity: Equity curve, shape [num_configs, num_bars]

    Returns:
        Max drawdowns (as positive decimals), shape [num_configs]
    """
    if equity.shape[1] == 0:
        return cp.zeros(equity.shape[0], dtype=cp.float32)

    # Running maximum (cumulative max along bar axis)
    running_peak = _running_max_gpu(equity)

    # Drawdown at each point: (peak - equity) / peak
    # Guard against division by zero
    safe_peak = cp.where(cp.abs(running_peak) < 1e-10, 1e-10, running_peak)
    drawdown = (running_peak - equity) / safe_peak

    # Max drawdown per config
    return cp.max(drawdown, axis=1).astype(cp.float32)


def calmar_gpu(
    equity: cp.ndarray,
    initial_cash: float,
    years: float,
) -> cp.ndarray:
    """
    Compute Calmar ratio for multiple equity curves.

    Calmar = CAGR / Max Drawdown
    Matches Rust implementation.

    Args:
        equity: Equity curve, shape [num_configs, num_bars]
        initial_cash: Starting capital
        years: Number of years in backtest period

    Returns:
        Calmar ratios, shape [num_configs]
    """
    cagr_vals = cagr_gpu(equity, initial_cash, years)
    max_dd = max_drawdown_gpu(equity)

    # Calmar = CAGR / max_drawdown (where max_drawdown > 0)
    calmar = cp.where(max_dd > 0, cagr_vals / max_dd, 0.0)

    return calmar.astype(cp.float32)


def win_rate_gpu(
    entry_bars: cp.ndarray,
    exit_bars: cp.ndarray,
    open_prices: cp.ndarray,
    fees_bps: float = 10.0,
    qty_per_trade: float = 100.0,
) -> cp.ndarray:
    """
    Compute win rate for multiple configs.

    Win rate = winning_trades / total_trades

    Args:
        entry_bars: Boolean array where entries occurred [num_configs, num_bars]
        exit_bars: Boolean array where exits occurred
        open_prices: Open prices, shape [num_bars]
        fees_bps: Fees in basis points per side
        qty_per_trade: Shares per trade

    Returns:
        Win rates, shape [num_configs]
    """
    num_configs, num_bars = entry_bars.shape
    fee_mult = fees_bps / 10000.0

    # Broadcast prices if needed
    if open_prices.ndim == 1:
        open_prices = cp.broadcast_to(open_prices, (num_configs, num_bars))

    # Count trades and winning trades
    entry_price = cp.zeros((num_configs,), dtype=cp.float32)
    num_trades = cp.zeros((num_configs,), dtype=cp.int32)
    num_wins = cp.zeros((num_configs,), dtype=cp.int32)

    for t in range(num_bars):
        # Record entry prices
        entering = entry_bars[:, t]
        entry_price = cp.where(entering, open_prices[:, t], entry_price)

        # Check wins on exits
        exiting = exit_bars[:, t]
        if cp.any(exiting):
            exit_price = open_prices[:, t]
            gross_pnl = (exit_price - entry_price) * qty_per_trade
            fees = (entry_price + exit_price) * qty_per_trade * fee_mult
            net_pnl = gross_pnl - fees

            is_win = net_pnl > 0
            num_wins = cp.where(exiting & is_win, num_wins + 1, num_wins)
            num_trades = cp.where(exiting, num_trades + 1, num_trades)

    # Win rate
    win_rate = cp.where(num_trades > 0, num_wins.astype(cp.float32) / num_trades, 0.0)

    return win_rate.astype(cp.float32)


def profit_factor_gpu(
    entry_bars: cp.ndarray,
    exit_bars: cp.ndarray,
    open_prices: cp.ndarray,
    fees_bps: float = 10.0,
    qty_per_trade: float = 100.0,
) -> tuple[cp.ndarray, cp.ndarray, cp.ndarray]:
    """
    Compute profit factor for multiple configs.

    Profit factor = gross_profit / gross_loss
    Matches Rust implementation.

    Args:
        entry_bars: Boolean array where entries occurred [num_configs, num_bars]
        exit_bars: Boolean array where exits occurred
        open_prices: Open prices, shape [num_bars]
        fees_bps: Fees in basis points per side
        qty_per_trade: Shares per trade

    Returns:
        Tuple of (profit_factor, gross_profit, gross_loss)
    """
    num_configs, num_bars = entry_bars.shape
    fee_mult = fees_bps / 10000.0

    # Broadcast prices if needed
    if open_prices.ndim == 1:
        open_prices = cp.broadcast_to(open_prices, (num_configs, num_bars))

    entry_price = cp.zeros((num_configs,), dtype=cp.float32)
    gross_profit = cp.zeros((num_configs,), dtype=cp.float32)
    gross_loss = cp.zeros((num_configs,), dtype=cp.float32)

    for t in range(num_bars):
        # Record entry prices
        entering = entry_bars[:, t]
        entry_price = cp.where(entering, open_prices[:, t], entry_price)

        # Compute PnL on exits
        exiting = exit_bars[:, t]
        if cp.any(exiting):
            exit_price = open_prices[:, t]
            gross_pnl = (exit_price - entry_price) * qty_per_trade
            fees = (entry_price + exit_price) * qty_per_trade * fee_mult
            net_pnl = gross_pnl - fees

            # Accumulate profits and losses
            is_profit = net_pnl > 0
            gross_profit = cp.where(
                exiting & is_profit,
                gross_profit + net_pnl,
                gross_profit,
            )
            gross_loss = cp.where(
                exiting & ~is_profit,
                gross_loss + cp.abs(net_pnl),
                gross_loss,
            )

    # Profit factor: gross_profit / gross_loss
    # Handle edge cases matching Rust
    pf = cp.where(
        gross_loss > 0,
        gross_profit / gross_loss,
        cp.where(gross_profit > 0, cp.inf, 0.0),
    )

    return pf.astype(cp.float32), gross_profit, gross_loss


def num_trades_gpu(exit_bars: cp.ndarray) -> cp.ndarray:
    """
    Count completed trades for multiple configs.

    A completed trade is an exit (matching Rust implementation).

    Args:
        exit_bars: Boolean array where exits occurred [num_configs, num_bars]

    Returns:
        Number of trades per config, shape [num_configs]
    """
    return cp.sum(exit_bars, axis=1).astype(cp.int32)


def turnover_gpu(
    entry_bars: cp.ndarray,
    exit_bars: cp.ndarray,
    open_prices: cp.ndarray,
    equity: cp.ndarray,
    initial_cash: float,
    years: float,
    qty_per_trade: float = 100.0,
) -> cp.ndarray:
    """
    Compute annual turnover for multiple configs.

    Turnover = (total traded notional / average capital) / years
    Matches Rust implementation.

    Args:
        entry_bars: Boolean array where entries occurred [num_configs, num_bars]
        exit_bars: Boolean array where exits occurred
        open_prices: Open prices, shape [num_bars]
        equity: Equity curve [num_configs, num_bars]
        initial_cash: Starting capital
        years: Number of years in backtest period
        qty_per_trade: Shares per trade

    Returns:
        Annual turnover per config, shape [num_configs]
    """
    num_configs, num_bars = entry_bars.shape

    # Broadcast prices if needed
    if open_prices.ndim == 1:
        open_prices = cp.broadcast_to(open_prices, (num_configs, num_bars))

    # Total traded notional
    entry_notional = cp.sum(
        cp.where(entry_bars, open_prices * qty_per_trade, 0.0),
        axis=1,
    )
    exit_notional = cp.sum(
        cp.where(exit_bars, open_prices * qty_per_trade, 0.0),
        axis=1,
    )
    total_traded = entry_notional + exit_notional

    # Average capital
    final_equity = equity[:, -1]
    avg_capital = (initial_cash + final_equity) / 2.0

    # Annual turnover
    turnover = cp.where(
        (years > 0) & (avg_capital > 0),
        (total_traded / avg_capital) / years,
        0.0,
    )

    return turnover.astype(cp.float32)


def compute_all_metrics_gpu(
    equity: cp.ndarray,
    entry_bars: cp.ndarray,
    exit_bars: cp.ndarray,
    open_prices: cp.ndarray,
    initial_cash: float = 100000.0,
    years: float = 10.0,
    fees_bps: float = 10.0,
    qty_per_trade: float = 100.0,
) -> dict[str, cp.ndarray]:
    """
    Compute all performance metrics for multiple configs.

    This is the main entry point for metrics computation.

    Args:
        equity: Equity curve [num_configs, num_bars]
        entry_bars: Boolean array where entries occurred
        exit_bars: Boolean array where exits occurred
        open_prices: Open prices, shape [num_bars]
        initial_cash: Starting capital
        years: Number of years in backtest period
        fees_bps: Fees in basis points per side
        qty_per_trade: Shares per trade

    Returns:
        Dict mapping metric names to arrays of shape [num_configs]
    """
    # Compute equity-based metrics
    total_ret = total_return_gpu(equity, initial_cash)
    cagr_val = cagr_gpu(equity, initial_cash, years)
    sharpe_val = sharpe_gpu(equity)
    sortino_val = sortino_gpu(equity)
    max_dd = max_drawdown_gpu(equity)
    calmar_val = calmar_gpu(equity, initial_cash, years)

    # Compute trade-based metrics
    win_rate_val = win_rate_gpu(entry_bars, exit_bars, open_prices, fees_bps, qty_per_trade)
    pf, gross_profit, gross_loss = profit_factor_gpu(
        entry_bars, exit_bars, open_prices, fees_bps, qty_per_trade
    )
    n_trades = num_trades_gpu(exit_bars)
    turn = turnover_gpu(
        entry_bars, exit_bars, open_prices, equity, initial_cash, years, qty_per_trade
    )

    return {
        "total_return": total_ret,
        "cagr": cagr_val,
        "sharpe": sharpe_val,
        "sortino": sortino_val,
        "max_drawdown": max_dd,
        "calmar": calmar_val,
        "win_rate": win_rate_val,
        "profit_factor": pf,
        "num_trades": n_trades,
        "turnover": turn,
    }


def metrics_to_numpy(metrics: dict[str, cp.ndarray]) -> dict[str, np.ndarray]:
    """
    Convert GPU metrics dict to numpy arrays for Polars/CPU consumption.

    Args:
        metrics: Dict from compute_all_metrics_gpu

    Returns:
        Dict with same keys but numpy arrays
    """
    return {k: cp.asnumpy(v) for k, v in metrics.items()}
